pkg/cloudprovider/nodecontroller/podevictor.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodecontroller
+
+import (
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/golang/glog"
+)
+
+type empty struct{}
+
+type PodEvictor struct {
+	lock         sync.Mutex
+	evictQueue   []string
+	evictedNodes map[string]empty
+
+	deletingPodsRateLimiter util.RateLimiter
+}
+
+func NewPodEvictor(deletingPodsRateLimiter util.RateLimiter) *PodEvictor {
+	return &PodEvictor{
+		evictQueue:              make([]string, 0),
+		evictedNodes:            make(map[string]empty),
+		deletingPodsRateLimiter: deletingPodsRateLimiter,
+	}
+}
+
+func (pe *PodEvictor) TryEvict(delFunc func(string)) error {
+	pe.lock.Lock()
+	defer pe.lock.Unlock()
+	for i, val := range pe.evictQueue {
+		if pe.deletingPodsRateLimiter.CanAccept() {
+			glog.Infof("Evicting pods on %s", val)
+			delFunc(val)
+			if i < len(pe.evictQueue)-1 {
+				pe.evictQueue = pe.evictQueue[i+1 : len(pe.evictQueue)]
+			} else {
+				pe.evictQueue = make([]string, 0)
+			}
+		} else {
+			break
+		}
+	}
+	return nil
+}
+
+func (pe *PodEvictor) AddNodeToEvict(nodeName string) {
+	pe.lock.Lock()
+	defer pe.lock.Unlock()
+	glog.Infof("Adding pods on %s", nodeName)
+
+	if _, ok := pe.evictedNodes[nodeName]; !ok {
```

Comment:
Can you reuse util.workqueue (you will have to pull the namespace/name of the node out on the other end, lookup its pods and evict them) or cache.fifo? I'm a little allergic to hand locking data structures :)